pkg/reconciler/delivery/delivery.go

```diff
@@ -99,53 +111,120 @@ func (c *Reconciler) updateRoute(ctx context.Context, cfg *v1.Configuration) err
 	if err != nil {
 		return err
 	}
+
 	logger.Info("Applying updated Route object")
 	_, err = c.client.ServingV1().Routes(cfg.Namespace).Update(route)
 
+	// when we have latestRevision = true, we know that we don't need to queue future events
+	if *route.Spec.Traffic[0].LatestRevision {
+		logger.Info("Progressive rollout completed!")
+	} else {
+		t, e := getThreshold(&policy, int(*route.Spec.Traffic[1].Percent))
+		if e != nil {
+			return e
+		}
+		logger.Infof("Queueing event for %v seconds later", t)
+		c.followup(cfg, time.Duration(t) * time.Second)
+	}
+
 	return err
 }
 
 // isRouteStatusUpToDate determines if the current Route status already matches our desired state
 func isRouteStatusUpToDate(route *v1.Route, newRevName string) bool {
-	for idx := range route.Status.Traffic {
-		if route.Status.Traffic[idx].RevisionName == newRevName {
-			return true
+	// the Route status is up to date if:
+	// 1. the new Revision is listed in the status traffic targets, AND
+	// 2. the status IngressReady condition's last transition time stamp hasn't expired
+	// OR if:
+	// 3. the new Revision is listed in the status traffic targets, AND
+	// 4. the new Revision already reached 100%
+	nameListed := false
+	for _, t := range route.Status.Traffic {
+		if t.RevisionName == newRevName {
+			nameListed = true
+			break
+		}
+	}
+	if !nameListed {
+		return false
+	}
+	if len(route.Status.Traffic) == 1 || *route.Status.Traffic[1].Percent == 100 {
+		return true
+	}
+	for _, c := range route.Status.Conditions {
+		if c.Type != "IngressReady" {
+			continue
+		}
+		if isLastTransitionExpired(c.LastTransitionTime.Inner, &policy, int(*route.Status.Traffic[1].Percent)) {
+			return false
 		}
 	}
 	return false
 }
 
+// isLastTransitionExpired determines if enough time has elapsed since the last transition of the IngressReady condition
+func isLastTransitionExpired(ltt metav1.Time, p *Policy, cp int) bool {
+	// this is the tricky part, using time stamps to figure out if Route is up to date
+	// IngressReady transition has a short delay of a couple seconds, which demands some wiggle room (the -5)
+	// this sacrifices some timing precision, but I haven't thought of a better way yet
+	t, e := getThreshold(&policy, cp)
+	// we can ignore error handling here, because returning true will cause a Route update
+	// modifyRouteSpec will discover the exact same error, and it can report that error more conveniently
+	if e != nil {
+		return true
+	}
+	return time.Now().After(ltt.Add(time.Duration(t - 5) * time.Second))
```

Comment:
You want this `5` to be a named constant with a comment, rather than here. It probably can be reused in the unit tests as well.

Reply:
I talked about comparing timestamps with Thomas in our 1:1 today, and I think there is a better way of comparing timestamps. I will implement it and submit a new version. It will get rid of `5`